Kubectl taint node based on label selector

ravisantoshgudimetla opened the pull request:
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes  #44522
**Release note**:
```
Taints the node based on label selector
```

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/44740)
<!-- Reviewable:end -->

jayunit100 commented:
Thanks @ravisantoshgudimetla this works for me :) 

```
cluster/kubectl.sh label nodes --all "myLabel=x" ; cluster/kubectl.sh taint nodes --selector='myLabel==x' dedicated=foo:PreferNoSchedule ;
node "127.0.0.1" labeled
node "127.0.0.1" tainted
```

Do we need the unit tests in order to merge this ? I'd prefer to push it forward since the unit tests (we think) don't seem to support combining kubectl labelling with tainting (although maybe there is way to do so, its just not obvious).

ravisantoshgudimetla commented:
/cc @jayunit100. 

- What is lacking?
Writing a uni test seems problematic, as the fake client is not able to build the information from labels.  

- What should we do?
So, the ideal way would be to have arguments to RunTaint(factory etc), fill in the values in test suite and fill in TaintOptions{} struct and build the information in test case. As this whole file needs some re-ordering(as discussed in #43171) the test-cases should be based on https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/taint.go#L362, if the node is tainted or not. The current approach seems to be checking from API object of node.(which should be in other file like TaintController etc).

ravisantoshgudimetla committed:
Changes to kubectl taint to respect selector flag

ravisantoshgudimetla commented:
cc @aveshagarwal

jayunit100 commented:
@k8s-bot test this please

jayunit100 commented:
cc @nimeshksingh

jayunit100 commented on pkg/kubectl/cmd/taint.go:
```diff
@@ -248,12 +254,12 @@ func (o *TaintOptions) Complete(f cmdutil.Factory, out io.Writer, cmd *cobra.Com
 	if o.all {
 		o.builder = o.builder.SelectAllParam(o.all).ResourceTypes("node")
 	} else {
-		if len(o.resources) < 2 {
```
@aveshagarwal ok to get rid of this check?

aveshagarwal replied:
It seems No. if both "selector" and "all"  are NOT provided, we should do this check to verify the sequence of <resource> <name>.

jayunit100 replied:
@aveshagarwal https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/taint.go#L236 isn't sufficient ?

ravisantoshgudimetla replied:
@aveshagarwal  So, in the below command, I have a resource `node` without selector and all options. Its throwing a valid error. 
`cluster/kubectl.sh taint node 127.0.0.1 `
`error: at least one taint update is required`

aveshagarwal replied:
I meant this errneous case:
kubectl  taint node key:value:Noschedule

aveshagarwal replied:
IOW, if you dont specify node name, but just the resource type node.

ravisantoshgudimetla replied:
Ok, thanks for the information @aveshagarwal. So, we already have a check at API server to handle it. There is no need to validate it here again.

cluster/kubectl.sh taint node dedicated=foo:PreferNoSchedule
error: You must provide one or more resources by argument or filename.
Example resource specifications include:
   '-f rsrc.yaml'
   '--filename=rsrc.json'
   'pods my-pod'
   'services'

aveshagarwal replied:
This is part of command line validation, and should be done in kubectl too.

ravisantoshgudimetla replied:
So, if that is the case, it should be something like if(o.selector)!=="" and o.all=false then we should throw this error. I can make that change.

aveshagarwal replied:
Sound good. And as you are at this, make sure one of `selector` and `all`  is present not both simultaneously if not already.

supereagle commented on pkg/kubectl/cmd/taint.go:
```diff
@@ -80,7 +80,13 @@ var (
+		kubectl taint nodes foo dedicated-
+
+		# Add from node 'foo' all the taints with key 'dedicated'
+		kubectl taint nodes foo dedicated-
```
The commands for `Add` and `Remove` examples are the same. Any error?

aveshagarwal replied:
It seems incorrect, the format with hyphen in the end ("dedicated-") is only for deletion. It does not make sense with addition.

ravisantoshgudimetla replied:
Right. Its a mistake, Thanks for catching it @supereagle & @aveshagarwal , I will change this.

kargakis commented:
@kubernetes/sig-cli-pr-reviews ptal

ravisantoshgudimetla commented:
@aveshagarwal PTAL.

aveshagarwal commented:
@ravisantoshgudimetla lgtm, now need to figure out the issue with your unit test.

jayunit100 commented:
@kubernetes/sig-cli-pr-reviews  Can we merge this, it fixes a pretty important label + kubectl bug  ? @ravisantoshgudimetla can you create a follow on issue for adding a unit test.

derekwaynecarr commented:
the code looks fine to me, for a unit test, we appear to have testing for label selectors in delete_test.go that may have a convention you could follow in `TestDeleteMultipleSelector`.  if there is no urgency, would prefer a test case with the code in one pr.

ravisantoshgudimetla commented:
@derekwaynecarr Thanks for the pointer regarding test case. I will check that and get back in case I am stuck.

ravisantoshgudimetla commented:
@derekwaynecarr I tried running the cmd using the way in deleted_test.go and it is causing the same error. The test case that is failing is https://gist.github.com/ravisantoshgudimetla/d66fb36b1279f118395e540cc8c61505#file-taint_test-go-L106.

As updated earlier, It seems the issue is with object being built using builder. The ones coming from unit tests are not having name created at https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/taint.go#L320. So, wondering if there is a way to work around that to write a unit  test. Initially I was thinking of mopdifying RunTaint() to take some arguments(especially for builders). Please lmk, if you have other thoughts or if I missed anything.

jayunit100 commented:
@derekwaynecarr  @ravisantoshgudimetla ok so sounds like to move forward, we have to do two things... 

1.  Lets create a follow on issue delineating the lack of the builder mocks needed for better higher level tests, and aim  to improve the general situation for `kubectl` tests in general, so that in the future we can more declaratively test the subtleties of taints w/ different builder+label+... situations and so on.  
2. per @derekwaynecarr suggestion , we can make a unit test for this patch by making a separate function for object creation.  

(derek let us know if this sounds good to you as well, i think that will solve everything brought up in this issue, including getting this much needed fix merged :))

derekwaynecarr commented:
i have no strong opinion on the style of the test, just would like a test of some kind.  i am not sure where we are these days with integration tests via kubectl but if that is easier than unit testing that is fine too i suspect.

ravisantoshgudimetla commented:
@derekwaynecarr @jayunit100, 
- I separated out the logic for flag validation and added a test case that validates it but for object builder mocking, I raised another issue.
- I have modified the existing test case to have a field called selector, so that it would be easier for extending later.

Please lmk, if you have any other concerns.

jayunit100 commented on pkg/kubectl/cmd/taint_test.go:
```diff
@@ -332,3 +334,50 @@ func TestTaint(t *testing.T) {
 	}
 }
+
+func TestValidateFlags(t *testing.T) {
```
yeah this is way cleaner, thanks ravi

jayunit100 commented on pkg/kubectl/cmd/taint_test.go:
```diff
@@ -332,3 +334,50 @@ func TestTaint(t *testing.T) {
+		},
+		{
+			taintOpts:   TaintOptions{selector: "", all: false, resources: []string{"node"}},
+			description: "With both selector and all flags and if node name is not provided",
```
fix desc

jayunit100 commented:
I think the test is sufficient to prevent regressions now... so i guess lets move this forward and follow up on #44953 to mock these out properly for better kubectl tests........  

PTAL @derekwaynecarr @aveshagarwal

aveshagarwal commented:
lgtm

ravisantoshgudimetla commented:
@jayunit100 - PTAL, I have addressed your comments on descriptions.

jayunit100 commented:
/lgtm

ravisantoshgudimetla commented:
@jayunit100 I did a gofmt again. Can you please ask the bot test?

ravisantoshgudimetla commented:
and the unit tests seem to be a flake..

jayunit100 commented:
@k8s-bot ok to test

ravisantoshgudimetla commented:
#45170  - Flake..

jayunit100 commented:
@k8s-bot unit test this please

jayunit100 commented:
/lgtm

jayunit100 commented:
@kubernetes/sig-cli-maintainers can we get an approve on this ? Thanks :)

pwittrock commented:
/approve
